File: c/euler.c
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <math.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include "euler.h"

int prime_sieve(int maxval, char **s, int **pp)
{
    int p, i, np = 0;
    int *primes = NULL;
    char *sieve = malloc(maxval);
    memset(sieve, 1, maxval);
    if (pp != NULL)
        primes = calloc(maxval, sizeof(int));

    sieve[0] = sieve[1] = 0;
    for (p = 0; p < maxval; p++)
    {
        if (sieve[p])
        {
            if (pp != NULL)
                primes[np++] = p;
            for (i = p * 2; i < maxval; i += p)
                sieve[i] = 0;
        }
    }

    *s = sieve;
    if (pp != NULL)
        *pp = realloc(primes, np * sizeof(int));
    return np;
}

char* nfactor_sieve(int maxval)
{
    int p, i;
    char *sieve = calloc(maxval, 1);
    for (p = 2; p < maxval; p++)
        if (sieve[p] == 0)
            for (i = p * 2; i < maxval; i += p)
                sieve[i]++;
    return sieve;
}

int is_prime(int n)
{
    int i, s = sqrt(n) + 1;
    if (n % 2 == 0 && n != 2)
        return 0;
    for (i = 3; i < s; i += 2)
        if (n % i == 0)
            return 0;
    return 1;
}

int count_factors(int n)
{
    int d, nf = 0;
    int s = sqrt(n);
    for (d = 1; d <= s; d++)
    {
        if (n % d == 0)
        {
            if (d * d == n)
                nf++;
            else
                nf += 2;
        }
    }
    return nf;
}

int all_factors(int n, int **ff)
{
    int i, d, nf;
    int s = sqrt(n);
    int *f = malloc(2 * s * sizeof(int));

    i = nf = 0;
    for (d = 1; d <= s; d++)
    {
        if (n % d == 0)
        {
            if (d * d == n)
            {
                nf++;
                f[i++] = d;
            }
            else
            {
                nf += 2;
                f[i++] = d;
                f[2 * s - i] = n / d;
            }
        }
    }

    memcpy(f + i, f + 2 * s - (nf / 2), (nf / 2) * sizeof(int));
    *ff = realloc(f, nf * sizeof(int));
    return nf;
}

int digit_usage(long long n, int d)
{
    int i = 0, v = 0;
    while (1)
    {
        if ((n == 0 && d == 0) || (d != 0 && i == d))
            break;
        v |= 1 << (n % 10);
        n /= 10;
        i++;
    }
    return v;
}

char* load_file(char *fname)
{
    int fd;
    char *data;
    struct stat st;

    fd = open(fname, O_RDONLY);
    fstat(fd, &st);
    data = malloc(st.st_size + 1);
    read(fd, data, st.st_size);
    data[st.st_size] = '\0';
    close(fd);
    return data;
}

double solve_quadratic(double a, double b, double c)
{
    double delta = sqrt(b * b - 4 * a * c);
    double r = (delta - b) / (2 * a);
    if (r < 0)
        r = (-delta - b) / (2 * a);
    return r;
}

int is_triangular(int t)
{
    double n = solve_quadratic(0.5, 0.5, -t);
    return (n == floor(n));
}

int is_pentagonal(int p)
{
    double n = solve_quadratic(1.5, -0.5, -p);
    return (n == floor(n));
}

File: py/euler.py
import math

product = lambda v: reduce(lambda x,y: x*y, v)

def prime_sieve(maxval):
    sieve = [True for n in xrange(maxval)]
    sieve[0] = sieve[1] = False
    for n in xrange(2, maxval):
        if sieve[n]:
            for x in xrange(n * 2, maxval, n):
                sieve[x] = False
    return sieve
def nfactor_sieve(maxval):
    sieve = [0 for n in xrange(maxval)]
    for n in xrange(2, maxval):
        if sieve[n] == 0:
            for x in xrange(n * 2, maxval, n):
                sieve[x] += 1
    return sieve

def is_prime(n, primes=None):
    if primes is not None:
        for p in primes:
            if n % p == 0:
                return False
        return True

    if n < 2:
        return False
    if n % 2 == 0 and n != 2:
        return False
    for d in xrange(3, int(math.sqrt(n)) + 1, 2):
        if n % d == 0:
            return False
    return True

def all_factors(n):
    f_lo = []
    f_hi = []
    s = math.sqrt(n)
    for d in xrange(1, int(s) + 1):
        if n % d == 0:
            if d * d == n:
                f_lo.append(d)
            else:
                f_lo.append(d)
                f_hi.insert(0, n/d)
    return f_lo + f_hi
count_factors = lambda n: len(all_factors(n))

load_word_file = lambda fname: [name[1:-1].lower() for name in file(fname).readlines()[0].split(',')]
word_score = lambda word: sum([ord(c) - ord('a') + 1 for c in word])

def digit_usage(n, d=None):
    v = 0
    if d is None:
        d = len(str(n))
    for i in xrange(d):
        v |= 1 << (n % 10)
        n /= 10
    return v

def reduce_fraction(n, d):
    dd = 2
    while dd < math.sqrt(d):
        while n % dd == 0 and d % dd == 0:
            n /= dd
            d /= dd
        dd += 1
    return (n, d)

def solve_quadratic(a, b, c):
    delta = math.sqrt(b * b - 4 * a * c)
    r = (delta - b) / (2 * a)
    if r < 0:
        r = (-delta - b) / (2 * a)
    return r
def is_triangular(t):
    r = solve_quadratic(0.5, 0.5, -t)
    return (r == math.floor(r))
def is_pentagonal(p):
    r = solve_quadratic(1.5, -0.5, -p)
    return (r == math.floor(r))
